docs/proposals/container-init.md

```diff
@@ -0,0 +1,460 @@
+computation prior to the start of the remaining containers. They may also, by
+virtue of their sequencing, block or delay the startup of application containers
+until some precondition is met. In this document we refer to the existing pod
+containers as **app containers**.
+
+This proposal also provides a high level design of **volume containers**, which
+initialize a particular volume, as a feature that specializes some of the tasks
+defined for init containers. The init container design anticipates the existence
+of volume containers and highlights where they will take future work
+
+## Design Points
+
+* Init containers should be able to:
+  * Perform initialization of shared volumes
+    * Download binaries that will be used in app containers as execution targets
+    * Inject configuration or extension capability to generic images at startup
+    * Perform complex templating of information available in the local environment
+    * Initialize a database by starting a temporary execution process and applying
+      schema info.
+  * Delay the startup of application containers until preconditions are met
+  * Register the pod with other components of the system
+* Reduce coupling:
+  * Between application images, eliminating the need to customize those images for
+    Kubernetes generally or specific roles
+  * Inside of images, by specializing which containers perform which tasks
+    (install git into init container, use filesystem contents
+    in web container)
+  * Between initialization steps, by supporting multiple sequential init containers
+* Init containers allow simple start preconditions to be implemented that are
+  decoupled from application code
+  * The order init containers start should be predictable and allow users to easily
+    reason about the startup of a container
+  * Complex ordering and failure will not be supported - all complex workflows can
+    if necessary be implemented inside of a single init container, and this proposal
+    aims to enable that ordering without adding undue complexity to the system.
+    Pods in general are not intended to support DAG workflows.
+* Both run-once and run-forever pods should be able to use init containers
+* As much as possible, an init container should behave like an app container
+  to reduce complexity for end users, for clients, and for divergent use cases.
+  An init container is a container with the minimum alterations to accomplish
+  its goal.
+* Volume containers should be able to:
+  * Perform initialization of a single volume
+  * Start in parallel
+  * Perform computation to initialize a volume, and delay start until that
+    volume is initialized successfully.
+  * Using a volume container that does not populate a volume to delay pod start
+    (in the absence of init containers) would be an abuse of the goal of volume
+    containers.
+* Container pre-start hooks are not sufficient for all initialization cases:
```

Comment:
Since pre-start hooks are not implemented yet, link to the issue: https://github.com/kubernetes/kubernetes/issues/140